pkg/gcp/token.go

```diff
@@ -1,57 +1,81 @@
+)
 
-func GetFederatedToken(projectNumber, pool, provider, token string) (string, error) {
-	ctx := context.Background()
-	requestData := getExchangeTokenRequestData(projectNumber, pool, provider, token)
+// GetFederatedToken tries to retrieve cached token from env variables, otherwise it will exchange
+// OIDC identity token to access token and cache them in env variables
+func GetFederatedToken(projectNumber, pool, provider, oidcToken string) (string, error) {
+	cachedToken := os.Getenv(gcpPubsubTokenKey)
+	cachedExpiresAt := os.Getenv(gcpPubsubTokenExpiryKey)
+	if tokenIsValid(cachedToken, cachedExpiresAt) {
+		log.Entry().Debug("reusing GCP PubSub access token from cache")
+		return cachedToken, nil
+	}
 
-	// data to byte
-	jsonData, err := json.Marshal(requestData)
+	ctx := context.Background()
+	token, expiresAt, err := exchangeOIDCToken(ctx, projectNumber, pool, provider, oidcToken)
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to marshal the request data")
+		return "", errors.Wrap(err, "token exchange")
+	}
+
+	os.Setenv(gcpPubsubTokenKey, token)
+	os.Setenv(gcpPubsubTokenExpiryKey, strconv.FormatInt(expiresAt, 10))
+	return token, nil
+}
+
+// exchangeOIDCToken exchanges OIDC identity token to access token and returns expiry time in Unix timestamp
+func exchangeOIDCToken(ctx context.Context, projectNumber, pool, provider, oidcToken string) (string, int64, error) {
+	if len(oidcToken) == 0 {
+		return "", 0, errors.New("OIDC identity token is absent")
 	}
 
-	// build request
-	request, err := http.NewRequestWithContext(ctx, http.MethodPost, exchangeTokenAPIURL, bytes.NewReader(jsonData))
+	stsService, err := sts.NewService(ctx, option.WithoutAuthentication())
 	if err != nil {
-		return "", errors.Wrap(err, "failed to build request")
+		return "", 0, errors.Wrap(err, "service not created")
 	}
 
-	// send request
-	response, err := http.DefaultClient.Do(request)
+	request := getExchangeTokenRequestData(projectNumber, pool, provider, oidcToken)
+	response, err := sts.NewV1Service(stsService).Token(request).Context(ctx).Do()
 	if err != nil {
-		return "", errors.Wrap(err, "failed to send request")
+		return "", 0, errors.Wrap(err, "exchange failed")
 	}
-	if response.StatusCode != http.StatusOK {
-		return "", fmt.Errorf("invalid status code: %v", response.StatusCode)
+
+	expiresAt := time.Now().Unix() + response.ExpiresIn
+	log.Entry().Debugf("token successfully exchanged and will expire at %s", time.Unix(expiresAt, 0))
+	return response.AccessToken, expiresAt, nil
+}
+
+func tokenIsValid(token string, expiresAtStr string) bool {
+	if token == "" {
+		return false
 	}
 
-	// response to data
-	defer response.Body.Close()
-	responseData := sts.GoogleIdentityStsV1ExchangeTokenResponse{}
-	err = json.NewDecoder(response.Body).Decode(&responseData)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to decode response")
+	expiresAt, _ := strconv.Atoi(expiresAtStr)
+	if int64(expiresAt) < time.Now().Unix() {
```

Comment:
There's a small chance that the token will be invalid between this validation and the actual use, maybe keep a small buffer to prevent this from happening?